PCR2 and 3 submission for review

jni2000 opened the pull request:
PCR.2 and PCR.3 for review

sbdt-google commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -0,0 +1,554 @@
+        ssl_key=config['sasTestHarnessKey'])
+
+    # Configure the Census Tract Directory that the PPA uses.
+    ppa.ConfigureCensusTractDriver(DEFAULT_CENSUSTRACTS_DATAPATH)
```
Useless line (already setup as default), and btw function do not exist anymore. All drivers moved to geo/drive.py

jni2000 replied:
This   line is no longer required due to refactor the code based on   geo/drive.py     This line has been removed now.

jni2000 commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -0,0 +1,554 @@
+  """
+
+  census_tract_driver = census_tract.CensusTractDriver()
+  census_tract_driver.SetCensusTractDirectory(DEFAULT_CENSUSTRACTS_DATAPATH)
```
Agreed.   We refactored the code  to align based   on the recent change in geo/drive.py

jni2000 commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -0,0 +1,554 @@
+# statements of any third-party software that are legally bundled with the
+# code in compliance with the conditions of those licenses.
+"""  Implementation of PCR tests """
+
```
Thanks   Serge that’s great . We have addressed this comments as the part of PCR 1   test     can you please review PCR2 and PCR3 alone in this PR. For all common   methods/functions you can still provide comments on PR480

jni2000 commented:
Serge, thanks for the review and comments. We have addressed them. Also, we have refactored  the PCR2 and PCR3 code based on PCR1 comments. Please  review only PCR2 and PCR3 tests alone in PR-502. If anything common please go ahead review in PCR 1 PR480

jni2000 commented:
Dear reviewers, we have updated this PR on top of the approved PCR.1, please review. Thanks.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -0,0 +1,643 @@
+    # the maximum PPA boundary created by SAS UUT shall be no more than 10% of the area
+    # of the maximum PPA boundary created by the Reference Model.
+    self.assertTrue(utils.PolygonsAlmostEqual(test_harness_ppa_geometry,
+                                              uut_ppa_geometry))
```
I'm assuming that everything above this line is unchanged in this PR. Can you please double-check + confirm?

Edit: I found and commented on a few changed lines. Please confirm that those are the only changes made above this line, or let me know which other ones to look at.

jni2000 replied:
Apart   from above two comments rework we have corrected few indentation fix to   maintain the uniformity across this file.     Line No: 221 -  4 space indentation   for line continuation     Line No: 247 -  4 space indentation   for line continuation

kate-harrison commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -0,0 +1,643 @@
+    # Write SAS UUT PPA to output directory of PCR.1 test.
+    # PPA Zone received from SAS UUT in PCR.1 test will be considered as input
+    # for PCR 3,PCR 6 and PCR 7 tests.
+    ppa_zone_data_file_path = os.path.join('testcases', 'output', 'test_WINNF_FT_S_PCR_1',
```
Will this auto-create the directory if it does not exist?

I think you need to include the config filename in this filename so that you don't overwrite the output from config 1 when you execute with config 2.

jni2000 replied:
Agreed   Added the check to ensure that directory is not exist then create and then   dump the ppa_zone data.. Config_file_name is now appended with   sas_uut_claimed_ppa_boundary-default.config.json

kate-harrison commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -0,0 +1,643 @@
+    # for PCR 3,PCR 6 and PCR 7 tests.
+    ppa_zone_data_file_path = os.path.join('testcases', 'output', 'test_WINNF_FT_S_PCR_1',
+                                           'sas_uut_claimed_ppa_boundary.json')
+    writeConfig(ppa_zone_data_file_path, uut_ppa_zone_data['zone'])
```
It's probably better to dump to JSON directly; this is a little confusing as-is.

jni2000 replied:
We used   the json API to directly dump data based on the suggestion

kate-harrison commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -0,0 +1,643 @@
+
+    # Create the actual config.
+    devices = [device_a, device_b, device_c]
+    config = {
```
I wonder if it is worth taking only two params in the config: (1) the filename for the PCR.1 config and (2) the provided contour. Then we can ensure that the params are exactly the same as those in PCR.1 and avoid the potential for misconfiguration errors. What do you think?

jni2000 replied:
Agreed .   Yes It is good suggestion. We will achieve better readability of testcase and   indicates clearly we have used the PCR 1 test configuration as input in this   test.     Modified the code accordingly.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -0,0 +1,643 @@
+    ppa_id = self.triggerPpaCreationAndWaitUntilComplete(ppa_creation_request)
+    logging.debug('ppa_id received from SAS UUT:%s', ppa_id)
+
+    # Notify SAS UUT about SAS Harness and trigger Full Activity Dump and
```
You haven't done the first part of this (notify).

jni2000 replied:
Notify   SAS UUT via injectPeer SAS admin API is already done inside the   method     triggerFadGenerationAndRetrievePpaZone . So avoid the confusion we have   repharse the     comment line accordingly

kate-harrison commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -0,0 +1,643 @@
+    ppa_id = self.triggerPpaCreationAndWaitUntilComplete(ppa_creation_request)
+    logging.debug('ppa_id received from SAS UUT:%s', ppa_id)
+
+    # Notify SAS UUT about SAS Harness and trigger Full Activity Dump and retrieves the
```
You haven't done the first part of this (notify).

jni2000 replied:
Notify SAS UUT via injectPeer SAS admin API is already done inside the method triggerFadGenerationAndRetrievePpaZone . So avoid the confusion we have repharse the comment line accordingly

kate-harrison commented on src/harness/testcases/output/test_WINNF_FT_S_PCR_1/sas_uut_claimed_ppa_boundary.json:
```diff
@@ -0,0 +1,1686 @@
+{
```
This probably shouldn't be checked in since it will be different for each SAS.

jni2000 replied:
This   file is no longer required to checkin based on the reason mentioned     so we have removed this file.

sbdt-google commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -0,0 +1,643 @@
+
+    # Set the values of fipsCode in pal_records_a and pal_records_b to make them adjacent.
+    # 20063955100 and 20063955200 respectively.
+    pal_record_a['fipsCode'] = 20063955100
```
FW has injected the census data in json format in the SAS-Data folder.

James, I dont see the scripts used to retrieve and process the census data in the src/data/ directory. As you know, any data that is used in SAS (DEM, land cover, zones, ...) is associated with a script creating the data. Hence we have reproducibility and perfect knowledge of what happened during this data creation.
Could you put the scripts as part of another PR there ?

jni2000 replied:
We will add the script.

sbdt-google commented:
Now that PCR1 is submitted, and since this PR has a lot of duplicated code, so difficult to review only difference, can you rebase your changes on the master (or merge somewhat), whatever is better so that we see only the new changes.

Thanks

jni2000 commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -0,0 +1,643 @@
+        'longitude'] = 38.72281, -100.50103
+
+    # Set the AntennaGain and EIRP capability.
+    device_a['installationParam']['eirpCapability'] = 30
```
In   order to highlight the below constraint           "The CBSD locations, antenna gains, and EIRP limits are chosen in a   way that only one PPA zone is created by those CBDSs using PPA Creation   Reference Model."          We have changed the default value here.

jni2000 commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -0,0 +1,643 @@
+                                        config['conditionalRegistrationData'])
+
+    # Trigger PPA Creation to SAS UUT.
+    test_harness_ppa_geometry = ppa.PpaCreationModel(config['registrationRequests'],
```
Agreed.   This required to be reflected in test spec as well. Can we track this changes   in separate PR once it is approved.

kate-harrison replied:
Good idea, Serge. We've been trying to match the spec step-for-step so far (though it isn't always possible) to aid with ITS' V&V, but I think that the benefits of deviating outweigh the risks here. I suggest we simply change the order within the code and add an explanation to the top of the file. If you agree, please either make the change in this PR or create an issue to track it.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -313,6 +313,20 @@ def test_WINNF_FT_S_PCR_1(self, config_filename):
 
+    # Write SAS UUT PPA to output directory of PCR.1 test.
+    # PPA Zone received from SAS UUT in PCR.1 test will be considered as input
+    # for PCR 3,PCR 6 and PCR 7 tests.
```
ITS will need to ensure that (maybe we can provide some tools to help?) since it is logically impossible to run PCR.3 before PCR.1.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -313,6 +313,20 @@ def test_WINNF_FT_S_PCR_1(self, config_filename):
+    # PPA Zone received from SAS UUT in PCR.1 test will be considered as input
+    # for PCR 3,PCR 6 and PCR 7 tests.
+    ppa_zone_data_dir_path = os.path.join('testcases', 'output', 'test_WINNF_FT_S_PCR_1')
+    ppa_zone_data_file_name = 'sas_uut_claimed_ppa_boundary_' + \
```
Consider pulling this out into its own function so that PCR.1 and PCR.3 do the same thing.

jni2000 replied:
Moved   these section to getSasUutClaimedPpaBoundaryFilePath function to return the   absolute path of PPA zone is stored that is received from SAS UUT as part of   PCR 1.

jni2000 commented:
Kate, we have responded your latest three review comments, please take a look before we merge the code. Thanks.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -296,6 +316,16 @@ def test_WINNF_FT_S_PCR_1(self, config_filename):
     cbsd_ids = self.assertRegistered(config['registrationRequests'],
                                      config['conditionalRegistrationData'])
 
+    # Asserts the REG-Conditional value doesn't exist in the registrationRequest,
```
Consider moving the PAL record injection *after* this block so that the ref model will catch PAL record errors (instead of the SAS UUT). It doesn't match the spec exactly but we are already deviating for the same reason.

jni2000 replied:
We   have already made necessary changes to reorder the way we trigger the PPA   reference model and capture any issue before trigger ppa creation admin api   to SAS UUT.     The way we have reordered mentioned as below:     1. assertRegConditionalsForPpaRefModel    method asserts and prepare the reqistration request by filling the   necessary  field required for PPA   reference model(This section has been already reviewed)     2. The below line will invoke ppa creation model before instruct SAS UUT to   trigger ppa boundary     ppa.PpaCreationModel(config['registrationRequests'],                                                            config['palRecords'])     3. Prepares ppa_creation_request for ppa_creation admin api and trigger PPA   creation to SAS UUT and waits for response and validate ppa creation at SAS   UUT is pass or fail based on     status check API . This will be accomplished by invoking the instance   method(This section has been already reviewed)         ppa_id =   self.triggerPpaCreationAndWaitUntilComplete(ppa_creation_request)          Let me know is there any change required re-order these API.?

kate-harrison replied:
My suggestion was that you move `ppa.PpaCreationModel(...)` to before `self._sas_admin.InjectPalDatabaseRecord(...)` so that we catch any errors with the PAL DB records in the reference model, not in the SAS UUT. This applies to PCR.1 and PCR.2, not to PCR.3.

kate-harrison replied:
Yes, I think it's OK to do assertRegistered() first. This looks good now. Thanks!

kate-harrison commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -510,18 +530,21 @@ def test_WINNF_FT_S_PCR_2(self, config_filename):
+                                               'default.config')
+    pcr_1_test_config_file_name = pcr_1_test_config_file_path.split('/')[-1]
+    sas_uut_claimed_ppa_boundary_file_path = getSasUutClaimedPpaBoundaryFilePath(
+        pcr_1_test_config_file_name)
```
Can't you just use "default.config" directly?

jni2000 replied:
Based on   the suggestion we have passed only "config_file_name" only to the   function     getSasUutClaimedPpaBoundaryFilePath()

kate-harrison commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -510,18 +530,21 @@ def test_WINNF_FT_S_PCR_2(self, config_filename):
         user_claimed_ppa_contour = json.load(claimed_ppa_file)
     except IOError:
-      raise RuntimeError('ConfigError:There is an error in reading path:%s'
+      raise RuntimeError('ConfigError:There is an error in reading path:%s \n'
```
This error message would be useful in the body of the test case as well (since ITS will not use our default config generation function to create their configs).

jni2000 replied:
Agreed.   Bassed on the suggestion, This error message to indicate execute PCR 1 test   first to proceed the dependency tests and validate user claimed contour path   exist moved to corresponding test case method rather than reside in configure   method.

jni2000 commented:
Kate, we have addressed your latest three review comments and made some changes. Please take a look before we proceed with the merge. Thanks.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -530,32 +531,16 @@ def test_WINNF_FT_S_PCR_2(self, config_filename):
     """Generate the WinnForum configuration for PCR 3."""
     # File path where SAS UUT claimed ppa boundary generated in PCR.1 test
+    pcr_1_test_config_file_name = 'default.config'
     pcr_1_test_config_file_path = os.path.join('testcases', 'configs',
```
Can this be removed?

jni2000 replied:
Extra   variable representing the 'default.config' has been removed.

kate-harrison replied:
I meant: are you using `pcr_1_test_config_file_path` anywhere? I think you can delete *that* variable.

janetliuboston replied:
I think pcr_1_test_config_file_path is being used in generate_PCR_3_default_config:
    pcr_1_test_config_file_path = os.path.join('testcases', 'configs',
                                               'test_WINNF_FT_S_PCR_1',
                                               'default.config')

    config = {
        'configPCR_1': pcr_1_test_config_file_path,
        'userClaimedPpaContour': user_claimed_ppa_contour_geometry
    }

and in test_WINNF_FT_S_PCR_3:
    pcr_1_test_config = loadConfig(config['configPCR_1'])

kate-harrison replied:
Oops, good point! Thank you :) Merge when ready!

kate-harrison commented on src/harness/testcases/WINNF_FT_S_PCR_testcase.py:
```diff
@@ -530,32 +531,16 @@ def test_WINNF_FT_S_PCR_2(self, config_filename):
     config = {
         'configPCR_1': pcr_1_test_config_file_path,
-        'userClaimedPpaContour': user_claimed_ppa_contour_geometry
+        'userClaimedPpaContourFilePath': sas_uut_claimed_ppa_boundary_file_path
```
This should take the provided contour itself, *not* the contour the came from PCR.1. ITS needs to be able to test with arbitrary contours, not just contours which are slightly smaller than the one created by the SAS UUT. Please undo this most recent change: your previous code was correct to use `'userClaimedPpaContour': user_claimed_ppa_contour_geometry`.

jni2000 replied:
We have   moved to assert the contour file is not found raise an exception for PCR 3   test. So repharse the error message and revert it back previous state

jni2000 commented:
Kate, please can you take a look of the changes we made to address your comments of April 16 ? Thanks
